contracts/pcv/utils/ERC20Allocator.sol

```diff
@@ -0,0 +1,454 @@
+        /// will be true if both are address 0, but then check drip condition will fail
+        /// meaning we do not have to check if address is 0 here
+        require(
+            pcvDepositToPSM[pcvDeposit] == psm,
+            "ERC20Allocator: invalid pcvDeposit"
+        );
+        _drip(psm, PCVDeposit(pcvDeposit));
+    }
+
+    /// helper function that does the dripping
+    function _drip(address psm, PCVDeposit pcvDeposit) internal {
+        /// Check
+        require(
+            _checkDripCondition(psm, pcvDeposit),
+            "ERC20Allocator: drip condition not met"
+        );
+
+        (uint256 amountToDrip, uint256 adjustedAmountToDrip) = getDripDetails(
+            psm,
+            pcvDeposit
+        );
+
+        /// Effects
+
+        /// deplete buffer with adjusted amount so that it gets
+        /// depleted uniformly across all assets and deposits
+        _depleteBuffer(adjustedAmountToDrip);
+
+        /// Interaction
+
+        /// drip amount to pcvDeposit psm so that it has targetBalance amount of tokens
+        pcvDeposit.withdraw(psm, amountToDrip);
+        emit Dripped(amountToDrip);
+    }
+
+    /// @notice does an action if any are available
+    /// @param pcvDeposit whose corresponding peg stability module action will be run on
+    function doAction(address pcvDeposit) external whenNotPaused {
+        address psm = pcvDepositToPSM[pcvDeposit];
+        if (_checkDripCondition(psm, PCVDeposit(pcvDeposit))) {
+            _drip(psm, PCVDeposit(pcvDeposit));
+        } else if (_checkSkimCondition(psm)) {
+            _skim(psm, pcvDeposit);
+        }
+    }
+
+    /// ----------- PURE & VIEW Only APIs -----------
+
+    /// @notice returns the target balance for a given PSM
+    function targetBalance(address psm) public view returns (uint256) {
+        return allPSMs[psm].targetBalance;
+    }
+
+    /// @notice function to get the adjusted amount out
+    /// @param amountToDrip the amount to adjust
+    /// @param decimalsNormalizer the amount of decimals to adjust amount by
+    function getAdjustedAmount(uint256 amountToDrip, int8 decimalsNormalizer)
+        public
+        pure
+        returns (uint256 adjustedAmountToDrip)
+    {
+        if (decimalsNormalizer == 0) {
+            adjustedAmountToDrip = amountToDrip;
+        } else if (decimalsNormalizer < 0) {
+            uint256 scalingFactor = 10**(-1 * decimalsNormalizer).toUint256();
+            adjustedAmountToDrip = amountToDrip / scalingFactor;
+        } else {
+            uint256 scalingFactor = 10**decimalsNormalizer.toUint256();
+            adjustedAmountToDrip = amountToDrip * scalingFactor;
+        }
+    }
+
+    /// @notice return the amount that can be skimmed off a given PSM
+    /// @param pcvDeposit pcv deposit whose corresponding psm will have skim amount checked
+    /// returns amount that can be skimmed, adjusted amount to skim and target to send proceeds
```

Comment:
add notice, this function only works if eligible for skim, otherwise will revert